From 513cb488bba41626807a046fb04ac054aef32094 Mon Sep 17 00:00:00 2001
From: Julian Greilich <j.greilich@gmx.de>
Date: Wed, 13 May 2026 12:04:23 +0200
Subject: a11y macOS: Respect state.expandable and state.expanded

We cannot implement the accessiblityExpanded property getter since that
would lead to every component being announced as collapsed, even when
expandable is false.

Task-number: QTBUG-146552
Change-Id: I81d5c35ea4f7eba61e63a4447d21bc743e01c140
---
 .../cocoa/qcocoaaccessibilityelement.mm       | 23 +++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git x/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibilityelement.mm y/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibilityelement.mm
index e517ac0044978e95bfe8a8895cff700055da90a9..0bc9e9b64c7436e22ba99ee74d4113027ad6b88a 100644
--- x/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibilityelement.mm
+++ y/qtbase/src/plugins/platforms/cocoa/qcocoaaccessibilityelement.mm
@@ -813,6 +813,29 @@ static void convertLineOffset(QAccessibleTextInterface *text, int *line, int *of
     return 0;
 }
 
+- (NSArray *)accessibilityAttributeNames {
+    NSMutableArray *attributes = [[NSMutableArray new] autorelease];
+
+    QAccessibleInterface *iface = self.qtInterface;
+    if (iface && iface->state().expandable)
+        [attributes addObject:NSAccessibilityExpandedAttribute];
+
+    return attributes;
+}
+
+- (id)accessibilityAttributeValue:(NSString *)attribute {
+    QAccessibleInterface *iface = self.qtInterface;
+    if (!iface) {
+        qWarning() << "Called attribute on invalid object: " << axid;
+        return nil;
+    }
+
+    if ([attribute isEqualToString:NSAccessibilityExpandedAttribute])
+        return iface->state().expanded ? @YES : @NO;
+
+    return nil;
+}
+
 - (NSArray *)accessibilityParameterizedAttributeNames {
 
     QAccessibleInterface *iface = self.qtInterface;
